projects/driver_controls/src/throttle.c

```diff
@@ -0,0 +1,153 @@
+// The throttle_get_position function simply reads the position from throttle_storage.
+#include "throttle.h"
+#include <string.h>
+#include "event_queue.h"
+#include "input_event.h"
+
+// The time period between every update of the pedal readings.
+#define THROTTLE_UPDATE_PERIOD_MS 50
+
+// Scales a reading to a measure out of 12 bits based on the given range.
+#define THROTTLE_SCALE_READING_TO_12_BITS(reading, upper_bound, lower_bound) \
+  ((1 << 12) * ((reading) - (lower_bound)) / ((upper_bound) - (lower_bound)))
+
+// The tolerance for verifying channel readings match their supposed relationship.
+// This tolerance is defined for the scaled-to-12bits measures.
+#define THROTTLE_CHANNEL_SCALED_READINGS_TOLERANCE 10
+
+// Returns true if the channels match their supposed relationship.
+static bool prv_channels_synched(ThrottleStorage *throttle_storage, int16_t reading_main,
+                                 int16_t reading_secondary) {
+  if (throttle_storage == NULL) {
+    return false;
+  }
+  int16_t fraction_main = THROTTLE_SCALE_READING_TO_12_BITS(
+      reading_main, throttle_storage->calibration_data.main_accel_thresh,
+      throttle_storage->calibration_data.main_bottom_thresh);
+
+  int16_t fraction_secondary = THROTTLE_SCALE_READING_TO_12_BITS(
+      reading_secondary, throttle_storage->calibration_data.secondary_accel_thresh,
+      throttle_storage->calibration_data.secondary_bottom_thresh);
+
+  return (abs(fraction_main - fraction_secondary)) < THROTTLE_CHANNEL_SCALED_READINGS_TOLERANCE;
+}
+
+// This callback is called whenever a conversion is done. It sets the flags
+// so that we know conversions are happening.
+static void prv_flag_update_callback(Ads1015Channel channel, void *context) {
+  if (context == NULL) {
+    return;
+  }
+  ThrottleStorage *throttle_storage = context;
+  throttle_storage->reading_updated_flag = true;
+  throttle_storage->reading_ok_flag = true;
+}
+
+// The periodic callback which checks if readings are up to date and channels are in synch.
+// If they are, the event corresponding to the pedals position is raised.
+// If not, a pedal timout event is raised.
+static void prv_raise_event_timer_callback(SoftTimerID timer_id, void *context) {
+  if (context == NULL) {
+    return;
+  }
+  ThrottleStorage *throttle_storage = context;
+
+  int16_t reading_main;
+  int16_t reading_secondary;
+  StatusCode status_code;
+  status_code = ads1015_read_raw(throttle_storage->pedal_ads1015_storage,
+                                 throttle_storage->channel_main, &reading_main);
+  if (!status_code) {
+    event_raise(INPUT_EVENT_PEDAL_TIMEOUT, 0);
+    return;
+  }
+  status_code = ads1015_read_raw(throttle_storage->pedal_ads1015_storage,
+                                 throttle_storage->channel_secondary, &reading_secondary);
+  if (!status_code) {
+    event_raise(INPUT_EVENT_PEDAL_TIMEOUT, 0);
+    return;
+  }
+
+  if (throttle_storage->reading_updated_flag &&
```

Comment:
I feel like this could be simpler. It seems like a lot of duplicated code with different parameters.